### PR TITLE
Add honeypot field

### PR DESCRIPTION
**Pull request by sunlight7777777**

### CODE REVIEW

File: demos/client-example/src/App.tsx [hunk 1]
```diff
@@ -59,6 +59,7 @@ function App(props: AppProps) {
         serverUrl: getServerUrl(),
         mongoAtlasUri: process.env.PROSOPO_MONGO_EVENTS_URI || '',
         devOnlyWatchEvents: process.env._DEV_ONLY_WATCH_EVENTS === 'true' || false,
+        honeypotFlag: false,
```
Comment (forgetso) [inline demos/client-example/src/App.tsx]: Please change to use `createHoneypot`

File: demos/cypress-shared/cypress/e2e/honeypot.cy.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        cy.intercept('/dummy').as('dummy')
+        cy.wrap(solutions).as('solutions')
+    })
+
+    it('BotD catch (no mocks, do not fill honeypot)', () => {
+        cy.visit(Cypress.env('default_page')).then(() => {
+            cy.get(checkboxClass).should('be.visible')
+            cy.clickIAmHuman().then(checkAndClickCaptchas)
```
Comment (forgetso) [inline demos/cypress-shared/cypress/e2e/honeypot.cy.ts]: Where is the test check?

e.g.

```
expect(something).to.be.equal
expect(something).to.be.checked 
etc
```

File: demos/cypress-shared/cypress/e2e/honeypot.cy.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        cy.stubBotdDetect()
+        cy.visit(Cypress.env('default_page')).then(() => {
+            cy.get(checkboxClass).should('be.visible')
+            cy.honeypotExists().then((isHoneypotExists) => {
+                cy.wait('@mockBotDetection')
+                if (isHoneypotExists) {
+                    cy.get('input#firstname').type('I am a bot', { force: true })
+                    cy.get(checkboxClass, { timeout: 12000 }).first().click()
+                    cy.checkHoneypot().then((isHoneypotFilled) => {
+                        if (isHoneypotFilled) {
```
Comment (forgetso) [inline demos/cypress-shared/cypress/e2e/honeypot.cy.ts]: Why is there an `if` statement here? This is [not recommended](https://docs.cypress.io/guides/core-concepts/conditional-testing).

Where is the test check?

e.g.

```
expect(something).to.be.equal
expect(something).to.be.checked 
etc
```

File: demos/cypress-shared/cypress/e2e/honeypot.cy.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+    it('Not caught by BotD or honeypot (mock BotD, do not fill honeypot)', () => {
+        cy.stubBotdDetect()
+        cy.visit(Cypress.env('default_page')).then(() => {
+            cy.get(checkboxClass).should('be.visible')
+            cy.honeypotExists().then((isHoneypotExists) => {
+                cy.wait('@mockBotDetection')
+                if (isHoneypotExists) {
+                    cy.checkHoneypot().then((isHoneypotFilled) => {
+                        if (isHoneypotFilled) {
```
Comment (forgetso) [inline demos/cypress-shared/cypress/e2e/honeypot.cy.ts]: Same comment as previous regarding `if` statement and lack of `expect`.

File: packages/honeypot/package.json [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        "prettier": "npx prettier . --check --no-error-on-unmatched-pattern --ignore-path ../../.eslintignore",
+        "prettier:fix": "npm run prettier -- --write",
+        "lint": "npm run eslint && npm run prettier",
+        "lint:fix": "npm run eslint:fix && npm run prettier:fix"
+    },
+    "browserslist": [
+        "> 0.5%, last 2 versions, not dead"
+    ],
+    "dependencies": {
+        "@fingerprintjs/botd": "^1.9.0",
```
Comment (forgetso) [inline packages/honeypot/package.json]: Why is fingerprint a dependency? It doesn't appear to be imported anywhere.

File: packages/honeypot/tsconfig.cjs.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
+{
+    "extends": "../../tsconfig.cjs.json",
+    "compilerOptions": {
+        "rootDir": "./src",
+        "outDir": "./dist/cjs",
+        "lib": ["es6", "dom"],
+        "jsxImportSource": "@emotion/react"
```
Comment (forgetso) [inline packages/honeypot/tsconfig.cjs.json]: Is this required? @emotion/react is not used in your code.

File: demos/cypress-shared/tsconfig.json [hunk 7]
```diff
@@ -14,6 +14,9 @@
         },
         {
             "path": "../../packages/util"
+        },
+        {
+            "path": "../../packages/procaptcha-frictionless"
```
Comment (forgetso) [inline demos/cypress-shared/tsconfig.json]: Is this required? You have no imports from @procaptcha-frictionless in the cypress code.

File: packages/honeypot/tsconfig.cjs.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    },
+    "include": ["src/index.tsx"],
+    "references": [
+        {
+            "path": "../util/tsconfig.cjs.json"
```
Comment (forgetso) [inline packages/honeypot/tsconfig.cjs.json]: Is this required? You have no imports from @prosopo/util in your package.

File: packages/procaptcha-frictionless/src/botDetection.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+// limitations under the License.
+import { load } from '@fingerprintjs/botd'
+
+export const botDetection = {
+    detectBot: async () => {
+        try {
+            const botd = await load()
+            const result = botd.detect()
+
+            console.log(JSON.stringify(result))
```
Comment (forgetso) [inline packages/procaptcha-frictionless/src/botDetection.ts]: Please remove debug

File: packages/types/src/config/config.ts [hunk 10]
```diff
@@ -146,6 +146,7 @@ export const ProsopoClientConfigSchema = ProsopoBasicConfigSchema.merge(
         solutionThreshold: number().positive().max(100).optional().default(80),
         dappName: string().optional().default('ProsopoClientDapp'),
         serverUrl: string().optional(),
+        honeypotFlag: boolean().optional().default(false),
```
Comment (forgetso) [inline packages/types/src/config/config.ts]: please change to `createHoneypot`

**Review by forgetso — CHANGES_REQUESTED** [review]
A few comments, mostly on dependencies and tests.

File: packages/procaptcha-frictionless/src/botDetection.ts (hunk 9 above)
Comment (sunlight7777777) [reply packages/procaptcha-frictionless/src/botDetection.ts]: Okay, I will remove these
Thanks

File: packages/types/src/config/config.ts (hunk 10 above)
Comment (sunlight7777777) [reply packages/types/src/config/config.ts]: Okay, I will change

File: packages/honeypot/tsconfig.cjs.json (hunk 8 above)
Comment (sunlight7777777) [reply packages/honeypot/tsconfig.cjs.json]: Sorry, I didn't removed it after test
I will remove these

File: packages/honeypot/tsconfig.cjs.json (hunk 6 above)
Comment (sunlight7777777) [reply packages/honeypot/tsconfig.cjs.json]: Sorry, I didn't removed it after test
I will remove these

File: demos/cypress-shared/tsconfig.json (hunk 7 above)
Comment (sunlight7777777) [reply demos/cypress-shared/tsconfig.json]: Sorry, I didn't removed it after test
I will remove these

File: packages/honeypot/package.json (hunk 5 above)
Comment (sunlight7777777) [reply packages/honeypot/package.json]: Sorry, I didn't removed it after test
I will remove these

File: demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    it('Not bot from BotD but caught by honeypot (mock BotD, fill honeypot)', () => {
+        cy.stubBotdDetect()
+        cy.visit(Cypress.env('default_page')).then(() => {
+            cy.get(checkboxClass).should('be.visible')
+            cy.wait('@mockBotDetection')
+            cy.window().its('__botDetected').should('be.false')
+            cy.get(honeypotSelector).type('I am a bot', { force: true })
+            cy.get(checkboxClass, { timeout: 12000 }).first().click()
+            cy.get(honeypotSelector).should('have.value', 'I am a bot')
+            cy.clickIAmHuman().then(checkAndClickCaptchas)
```
Comment (forgetso) [inline demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts:68]: What is the expected outcome?

Comment (forgetso) [inline demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts:72]: What is the expected outcome?

File: demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    })
+
+    it('Not caught by BotD or honeypot (mock BotD, do not fill honeypot)', () => {
+        cy.stubBotdDetect()
+        cy.visit(Cypress.env('default_page')).then(() => {
+            cy.get(checkboxClass).should('be.visible')
+            cy.wait('@mockBotDetection')
+            cy.window().its('__botDetected').should('be.false')
+            cy.get(honeypotSelector).should('have.value', '')
+            cy.get(checkboxClass, { timeout: 12000 }).first().click()
```
Comment (forgetso) [inline demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts:85]: What is the expected outcome?

File: demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        cy.intercept('/dummy').as('dummy')
+        cy.wrap(solutions).as('solutions')
+    })
+
+    it('BotD catch (no mocks, do not fill honeypot)', () => {
+        cy.visit(Cypress.env('default_page')).then(() => {
+            cy.get(checkboxClass).should('be.visible')
+            cy.get(honeypotSelector).should('have.value', '')
+            cy.window().its('__botDetected').should('be.true')
+            cy.clickIAmHuman().then(checkAndClickCaptchas)
```
Comment (forgetso) [inline demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts:57]: What is the expected outcome?

File: demos/client-frictionless-example/src/App.tsx [hunk 14]
```diff
@@ -29,6 +29,7 @@ function App() {
             (process.env.PROSOPO_DEFAULT_ENVIRONMENT as EnvironmentTypes) || EnvironmentTypesSchema.enum.development,
         serverUrl: process.env.PROSOPO_SERVER_URL || '',
         atlasUri: process.env._DEV_ONLY_WATCH_EVENTS === 'true' || false,
+        createHoneypot: process.env.PROSOPO_HONEYPOT === 'true',
```
Comment (forgetso) [inline demos/client-frictionless-example/src/App.tsx]: Why is this here if there is a `getHoneypot` function?

File: demos/client-example/env.development [hunk 15]
```diff
@@ -7,4 +7,5 @@ PROSOPO_PORT=9230
 PROSOPO_DEFAULT_NETWORK=development
 PROSOPO_DEFAULT_ENVIRONMENT=development
 PROSOPO_MONGO_EVENTS_URI=mongodb+srv://<MONGO_URI_HERE>/frictionless_events
-_DEV_ONLY_WATCH_EVENTS=false
\ No newline at end of file
+_DEV_ONLY_WATCH_EVENTS=false
+PROSOPO_HONEYPOT=true
```
Comment (forgetso) [inline demos/client-example/env.development:11]: How will this work in github CICD workflow? The honeypot will always be enabled.

File: demos/cypress-shared/package.json [hunk 16]
```diff
@@ -27,9 +27,9 @@
         "build": "tsc",
         "clean": "tsc --build --clean",
         "cypress:open:client-example": "CYPRESS_BASE_URL='http://0.0.0.0:9230' cypress open",
-        "cypress:run:client-example": "CYPRESS_BASE_URL='http://0.0.0.0:9230' cypress run",
+        "cypress:run:client-example": "CYPRESS_BASE_URL='http://0.0.0.0:9230' cypress run --spec 'cypress/e2e/captcha.cy.ts,cypress/e2e/correct.captcha.cy.ts,cypress/e2e/correct.captcha.signup.cy.ts'",
         "cypress:open:client-example:frictionless": "CYPRESS_BASE_URL='http://0.0.0.0:9230' cypress open --env default_page='/frictionless'",
-        "cypress:run:client-example:frictionless": "CYPRESS_BASE_URL='http://0.0.0.0:9230' cypress run --env default_page='/frictionless'",
+        "cypress:run:client-example:frictionless": "CYPRESS_BASE_URL='http://0.0.0.0:9230' cypress run --env default_page='/frictionless' --spec 'cypress/e2e/honeypot-enabled.cy.ts'",
```
Comment (forgetso) [inline demos/cypress-shared/package.json]: Create an additional command `cypress:run:client-example:frictionless:honeypot` and specify the PROSOPO_HONEYPOT var on the command line. Add the additional command to the github workflows.

File: packages/procaptcha-frictionless/src/types/globals.d.ts [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+export {}
+
+declare global {
+    interface Window {
+        __botDetected: boolean
```
Comment (forgetso) [inline packages/procaptcha-frictionless/src/types/globals.d.ts]: This doesn't feel like something we should have in a public package. Remove this file.

Change the `onHuman` function in the demo code so that it sets `__botDetected`
```typescript
const callbacks = { onHuman: function() {//set the window variable } }
<ProcaptchaFrictionless config={config} callbacks={callbacks} />
```

**Review by forgetso — CHANGES_REQUESTED** [review]
Getting there. Some changes required to do with window variables, env, and workflow.

File: demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts (hunk 11 above)
Comment (sunlight7777777) [reply demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts:72]: cy.get("input[type='checkbox']").first().should('be.checked')
This is the result

File: demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts (hunk 12 above)
Comment (sunlight7777777) [reply demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts:85]: cy.get("input[type='checkbox']").first().should('be.checked')
This is the result (I will add for this condition)

File: demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts (hunk 13 above)
Comment (sunlight7777777) [reply demos/cypress-shared/cypress/e2e/honeypot-enabled.cy.ts:57]: cy.get("input[type='checkbox']").first().should('be.checked')
This is the result

File: packages/procaptcha-frictionless/src/types/globals.d.ts (hunk 17 above)
Comment (sunlight7777777) [reply packages/procaptcha-frictionless/src/types/globals.d.ts]: In fact, I didn't think about it
I didn't care about onHuman function
I will check this function and try to set  __botDetected in onHuman

File: demos/client-frictionless-example/src/App.tsx (hunk 14 above)
Comment (sunlight7777777) [reply demos/client-frictionless-example/src/App.tsx]: Okay, I will check this again and if it doesn't need, I will remove this

File: demos/cypress-shared/package.json (hunk 16 above)
Comment (sunlight7777777) [reply demos/cypress-shared/package.json]: Okay, I will try for this

File: demos/client-example/env.development (hunk 15 above)
Comment (sunlight7777777) [reply demos/client-example/env.development:11]: I solved this
